Dockerfile

```diff
@@ -1,76 +1,174 @@
 # Build stage
-FROM docker.io/library/golang:1.21-alpine3.18 AS build-env
+FROM docker.io/library/node:20-alpine3.18 AS build-frontend
+
+# Build deps
+RUN apk --no-cache add build-base git \
+    && rm -rf /var/cache/apk/*
+
+# Setup repo
+WORKDIR /usr/src/code.gitea.io/gitea
+
+COPY Makefile .
+
+# Download NPM Packages
+COPY package.json .
+COPY package-lock.json .
+
+RUN make deps-frontend
+
+# Copy source files
+COPY ./webpack.config.js .
+COPY ./assets ./assets
+COPY ./public ./public
+COPY ./web_src ./web_src
+
+ARG GITHUB_REF_NAME
+ARG GITHUB_REF_TYPE
+ARG DOCKER_GITEA_VERSION
+
+ENV GITHUB_REF_NAME=${GITHUB_REF_NAME:-docker-develop}
+ENV GITHUB_REF_TYPE=${GITHUB_REF_TYPE:-branch}
+ENV DOCKER_GITEA_VERSION=${DOCKER_GITEA_VERSION}
+
+# Build frontend
+RUN make clean-all frontend
+
+# Build stage
+FROM docker.io/library/golang:1.21-alpine3.18 AS build-backend
 
 ARG GOPROXY
 ENV GOPROXY ${GOPROXY:-direct}
 
-ARG GITEA_VERSION
 ARG TAGS="sqlite sqlite_unlock_notify"
 ENV TAGS "bindata timetzdata $TAGS"
 ARG CGO_EXTRA_CFLAGS
 
 # Build deps
-RUN apk --no-cache add \
-    build-base \
-    git \
-    nodejs \
-    npm \
+RUN apk --no-cache add build-base git \
     && rm -rf /var/cache/apk/*
 
 # Setup repo
-COPY . ${GOPATH}/src/code.gitea.io/gitea
 WORKDIR ${GOPATH}/src/code.gitea.io/gitea
 
-# Checkout version if set
-RUN if [ -n "${GITEA_VERSION}" ]; then git checkout "${GITEA_VERSION}"; fi \
- && make clean-all build
+COPY Makefile .
+
+# Download Golang Modules
+COPY go.mod .
+COPY go.sum .
+
+RUN make deps-backend
+
+# Copy source files
+COPY ./build ./build
+COPY ./cmd ./cmd
+COPY ./models ./models
+COPY ./modules ./modules
+COPY ./options ./options
+COPY ./routers ./routers
+COPY ./services ./services
+COPY ./templates ./templates
+COPY ./build.go .
+COPY ./main.go .
+
+# Clean directory
+RUN make clean-all
+
+# Copy frontend build artifacts
+COPY --from=build-frontend /usr/src/code.gitea.io/gitea/public ./public
+
+ARG GITHUB_REF_NAME
+ARG GITHUB_REF_TYPE
+ARG DOCKER_GITEA_VERSION
+
+ENV GITHUB_REF_NAME=${GITHUB_REF_NAME:-docker-develop}
+ENV GITHUB_REF_TYPE=${GITHUB_REF_TYPE:-branch}
+ENV DOCKER_GITEA_VERSION=${DOCKER_GITEA_VERSION-${GITHUB_REF_NAME}}
```

Comment:
This is the main change to highlight from the last couple commits.

We are forwarding the information provided in the GitHub Action envrionment to the docker build so that it can be utilised in the Makefile in place of the `.git/` directory.

For tags alone, the Makefile could have remained untouched. However, due to the way the nightly version is constructed, the `git describe` result has to be passed in and handled seperately to the currenlty available mechanisms.